### ISSUE TITLE
FR: Expose results from HTML5 Geolocation API to AMP

### ISSUE THREAD

**Issue by ericlindley-g** [ampproject/amphtml#8929]
Expose results from HTML5 Geolocation API to AMP: e.g. for bindings to amp-list[src], once the user has consented, give developers the ability to send lat/long information along with the server call so as to provide personalized results in response, based on geo

/cc @aghassemi   @choumx

**Comment by aghassemi** [comment]
One idea is to expose these through [variable substitution](https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md):  e.g. `GEO_LAT`, `GEO_LON`

Since these values are not always available, values could be empty. First time AMP runtime sees these variables, that's when user will be prompted for granting permission (variable substitution supports async values, so we don't need to wait for another page load after permission is granted to do substitution) 

@dvoytenko thoughts?
@choumx (somewhat related): do you have plans to expose our existing variables to bind? Would be nice to have access to them as amp-bind variables as well for greater flexibility.

**Comment by dvoytenko** [comment]
The user permission seems a pretty drastic ask to personalize a generic `amp-list`? Typically this would be done as a coarse fidelity via IP address.

Technically, some issues we'd run with when AMP is embedded in an iframe - which is faire amount of impressions. The only way would be to ask the viewer I believe, but I don't see how the viewer can pass it on to the document properly.

**Comment by ericlindley-g** [comment]
For context: the rationale for this request was because IP address wasn't offering enough fidelity for their purposes.

That's a good point about the viewer — does sound like a pretty big hurdle

**Comment by dvoytenko** [comment]
If we were to proceed, I can imagine we could have an extension that would request geo permissions when the page is executed on the direct origin - this could be useful for AMP as canonical and other cases. But we'd disable this extension when executed from proxy. Which means, GEO_ vars would also have to "understand" that they may not be available.

**Comment by aghassemi** [comment]
@dvoytenko Geo location permission can be requested by iframes (tested with Android and iOS) so I don't believe we need viewer integration for this (which is a good thing considering with CURLS user is giving the permission to a single publisher and not everyone in AMP)

**Comment by dvoytenko** [comment]
@aghassemi There are two problems with this:

1. The origin requesting permission will be that of proxy cache. That might not be clear for a user.
2. I believe both browsers are considering deprecating geo location in 3p iframes.

**Comment by dandv** [comment]
@ericlindley-g: just wanted to link to my previous issue, #826. If you can update [your comment](https://github.com/ampproject/amphtml/issues/8929#issuecomment-296909138) to link to it, I can delete this one.

**Comment by ericlindley-g** [comment]
@dandv  Updated — thanks!

**Comment by shishirm** [comment]
Hello. Just bumping this up again to see if this can be prioritized. As a larger corpus of AMP pages start to be built, this feature seems like a requirement to ensure continued adoption, especially as the pages built are deeper into the funnel. IP Based geolocation is limited in its accuracy and this approach  puts the control back in the hands of the user/creator.

Also, is the iframe based approach still viable i.e. will the iframe trigger a request to the user asking for permissions when the page is rendered from the amp cache? This is less desirable given the limitations of amp-iframe but this may be the only option until this feature is built out?

I know that the amp-geo component does give country level detection. Could that functionality be extended to provide more granular geolocation information?

**Comment by ericlindley-g** [comment]
Thanks @shishirm — it's unclear still if the Geolocation API is exactly the method that should be used for these use cases, but we are doing a deeper dive into geo- (and other types of) personalization this quarter, and would love to hear some specific use cases that we can use to develop solutions. Please feel free to share

**Comment by eduardogoncalves** [comment]
Can we check at load-time if user has been prompted to allow/deny geolocation?
```
navigator.permissions.query({'name': 'geolocation'})
.then( permission => console.log(permission) )
```
that can show:
```
>>> PermissionStatus {state: "prompt", onchange: null}
>>> PermissionStatus {state: "granted", onchange: null}
>>> PermissionStatus {state: "denied", onchange: null}
```

- If user has not been prompted yet, we wait until user taps a button;
- if user has already allowed we get lat/log;
- if user has denied we bind this info so the component that needs lat/log can use it's defaults.

**Comment by aghassemi** [comment]
@eduardogoncalves load-time is instant in AMP due to Server-side rendering, without making an extension render-blocking, by the time we get to `navigator.permissions.query`, page is already shown.

**Comment by jpettitt** [comment]
I did look very briefly at using the geolocation API in `amp-geo`. The reasons I didn't use it include:

1. For the specific use case, country level customizations for privacy permissions, asking for privacy centric data (precise location) would have been counter productive.
2. The need to prompt the user and the questions around the scope of the permission. For example if an iframed AMP page requests and received geolocation permissions does that permission carry over other pages iframed by the same origin which would be undesirable.
3. Undesirable render blocking while waiting for permission
4. Ambiguity about who was requesting permission

**Comment by eduardogoncalves** [comment]
Hi, any news on "expose results from HTML5 Geolocation API"?  I think the last discussion on this was on https://github.com/ampproject/amphtml/issues/18658#issuecomment-434853467

Will we create a new component or use amp-geo? Or we won't support getting user's current position provided by HTML5 Geolocation API ```window.navigator.geolocation.getCurrentPosition()```?

**Comment by aghassemi** [comment]
Hi,

Sorry for the late reply. We did a design review on this and a strong requirement was only allowing this on user-interaction. We don't want the prompt on page load.

We can have this feature by just adding a new action to amp-geo called getPreciseLocation which would trigger an event on amp-geo ( preciseLocation). When data is available. Event would have the lat and lon. No need for amp-state integration as user can setState as result of the event.

**Comment by safwans** [comment]
I'd love to learn more about this feature request. Is there more information available estimated launch date for getPreciseLocation?

**Comment by aghassemi** [comment]
Hi @safwans 

@cvializ has started working on this. Plan to launch by end of June.

We would love to hear your use-case and how you imagine to use this feature if you don't mind sharing. /cc @jeffjose

**Comment by safwans** [comment]
That's great news!

The use case in point is search results based on current location. With this feature, our partners can suggest listings that are closer to the user's current location.

**Comment by jeffjose** [comment]
Thanks for the extra context, @safwans. 

I think your usecase would be covered in the work that @cvializ is doing.

**Comment by aghassemi** [comment]
this has been implemented as `amp-user-location` see https://amp.dev/documentation/components/amp-user-location for details.

**Comment by caiotomazelli** [comment]
This is awesome, thank you for your work on this long-awaited feature!
Any timeline on moving this out of experimental? I got more than a handful of partners waiting eagerly to use it!

**Comment by aghassemi** [comment]
Credit goes to @cvializ , I just closed the issue 😆 . @cvializ can tell you about the timeline as well.

**Comment by cvializ** [comment]
We're still reviewing it to make sure it's a great experience, so stay tuned while we work on that!

**Comment by aaron-renovationbrands** [comment]
We are tuned in!

**Comment by kdnanmaga** [comment]
> this has been implemented as `amp-user-location` see https://amp.dev/documentation/components/amp-user-location for details.

This page is giving me a 404, is there any way to get the user's precise GPS coordinates? My use-case is a location based search.

**Comment by isaac-tru** [comment]
> > this has been implemented as `amp-user-location` see https://amp.dev/documentation/components/amp-user-location for details.
> 
> This page is giving me a 404, is there any way to get the user's precise GPS coordinates? My use-case is a location based search.

I am also getting a 404, any update as to where this has moved to? All I am getting is amp-geo.

**Comment by jeffjose** [comment]
We deprecated `amp-user-location` due to lack of usage. See https://github.com/ampproject/amphtml/issues/25573 for details.
